scheduling/taskmetadata.go

```
@@ -0,0 +1,463 @@
+
+	for _, run := range taskRuns.Runs[1:] {
+		if run.RunID > highestRunID {
+			highestRunID = run.RunID
+		}
+	}
+
+	return highestRunID, nil
+}
+
+func (m *TaskMetadataManager) GetRunByID(taskID TaskID, runID TaskRunID) (TaskRunMetadata, error) {
+	taskRunKey := TaskRunKey{taskID: taskID}
+	taskRunMetadata, err := m.storer.Get(taskRunKey.String())
+	if err != nil {
+		return TaskRunMetadata{}, fmt.Errorf("failed to fetch task: %v", err)
+	}
+
+	runs := TaskRuns{}
+	err = runs.Unmarshal([]byte(taskRunMetadata))
+	if err != nil {
+		return TaskRunMetadata{}, err
+	}
+
+	// Want to move this logic out
+	found := false
+	var runRecord TaskRunSimple
+	for _, run := range runs.Runs {
+		if run.RunID == runID {
+			runRecord = run
+			found = true
+			break
+		}
+	}
+	if !found {
+		return TaskRunMetadata{}, fmt.Errorf("run not found")
+	}
+
+	date := runRecord.DateCreated
+	taskRunMetadataKey := TaskRunMetadataKey{taskID: taskID, runID: runRecord.RunID, date: date}
+	rec, err := m.storer.Get(taskRunMetadataKey.String())
+	if err != nil {
+		return TaskRunMetadata{}, err
+	}
+
+	taskRun := TaskRunMetadata{}
+	err = taskRun.Unmarshal([]byte(rec))
+	if err != nil {
+		return TaskRunMetadata{}, fmt.Errorf("failed to unmarshal run record: %v", err)
+	}
+	return taskRun, nil
+}
+
+func (m *TaskMetadataManager) GetRunsByDate(start time.Time, end time.Time) (TaskRunList, error) {
+	// the date range is inclusive
+	var runs []TaskRunMetadata
+
+	// iterate through each day in the date range including the end date
+	for date := start; date.Before(end) || date.Equal(end); date = date.AddDate(0, 0, 1) {
+		dayRuns, err := m.getRunsForDay(date, start, end)
+		if err != nil {
+			return []TaskRunMetadata{}, err
+		}
+		runs = append(runs, dayRuns...)
+	}
+
+	return runs, nil
+}
+
+func (m *TaskMetadataManager) getRunsForDay(date time.Time, start time.Time, end time.Time) ([]TaskRunMetadata, error) {
+	// TODO: question, should this return an error if we can't marshal
```

imo we should throw an error. a faulty record could cause chaos downstream.